client/play-client/src/main/scala/sttp/tapir/client/play/EndpointToPlayClient.scala

```diff
@@ -95,16 +94,26 @@ private[play] class EndpointToPlayClient(clientOptions: PlayClientOptions, ws: S
           case EndpointIO.FixedHeader(_, codec, _)         => codec.decode(())
           case EndpointIO.Empty(codec, _)                  => codec.decode(())
           case EndpointOutput.OneOf(mappings, codec) =>
-            mappings
-              .find(mapping => mapping.statusCode.isEmpty || mapping.statusCode.contains(code)) match {
-              case Some(mapping) =>
-                getOutputParams(mapping.output, body, headers, code, statusText).flatMap(p => codec.decode(p.asAny))
-              case None =>
-                DecodeResult.Error(
-                  statusText,
-                  new IllegalArgumentException(s"Cannot find mapping for status code ${code} in outputs $output")
-                )
-            }
+            headers
+              .get(HeaderNames.ContentType)
+              .map { h => MediaType.parse(h.head) }
+              .map {
+                case Right(content) =>
+                  mappings collectFirst {
+                    case m if (m.statusCode.isEmpty || m.statusCode.contains(code)) && m.output.matchesContent(content) => m
+                  } match {
+                    case Some(mapping) =>
+                      getOutputParams(mapping.output, body, headers, code, statusText).flatMap(p => codec.decode(p.asAny))
+                    case None =>
+                      DecodeResult.Error(
+                        statusText,
+                        new IllegalArgumentException(s"Cannot find mapping for status code $code in outputs $output")
+                      )
+                  }
+                case Left(_) =>
+                  DecodeResult.Error(statusText, new IllegalArgumentException("Unable to parse Content-Type header"))
+              }
+              .getOrElse(DecodeResult.Error(statusText, new IllegalStateException("Missing Content-Type header")))
```

Comment:
what if one of the response variants has no body (e.g. a redirect or some other non-200 status code)?

Reply:
right, it has to be handled, adjusted